### PR TITLE
Sentinel - SourceControls 2023-06-01-preview

### PR DESCRIPTION
**Pull request by oformaniuk**
### Changelog
Add a changelog entry for this PR by answering the following questions:
1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
Service changes are ready and can be deployed as soon as Swagger changes are approved.

4. When do you expect to publish the swagger? Please provide date or, the date is not yet available, the month.
2023-06-01
6. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
  
- [x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version). Note that this doesn't apply if you are trying to merge a PR that was previously in the private repository.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/overview_breakingchanges).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](http://aka.ms/aprwiki))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 1]
```diff
@@ -213,18 +226,20 @@
             }
           }
         }
-      },
-      "put": {
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.OperationalInsights/workspaces/{workspaceName}/providers/Microsoft.SecurityInsights/sourcecontrols/{sourceControlId}/delete": {
+      "post": {
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:232]: >post [](http://example.com/codeflow?start=7&length=4)

Delete operation as POST?

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 2]
```diff
@@ -404,8 +417,41 @@
...
+        "kind": {
+          "description": "The kind of repository access credentials",
+          "$ref": "#/definitions/RepositoryAccessKind"
+        },
+        "code": {
+          "type": "string",
+          "description": "OAuth Code. Required when `kind` is `OAuth`"
+        },
+        "state": {
+          "type": "string",
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:432]: >string [](http://example.com/codeflow?start=19&length=6)

Should this be an enum?

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 3]
```diff
@@ -404,8 +417,41 @@
...
+          "description": "OAuth State. Required when `kind` is `OAuth`"
+        },
+        "clientId": {
+          "type": "string",
+          "description": "OAuth ClientId. Required when `kind` is `OAuth`"
+        },
+        "token": {
+          "type": "string",
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:440]: Use [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret) to mark the secret.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 4]
```diff
@@ -461,15 +521,18 @@
...
         "webhookUrl": {
           "description": "URL that gets invoked by the webhook.",
-          "type": "string"
+          "type": "string",
+          "readOnly": true
         },
         "webhookSecretUpdateTime": {
           "description": "Time when the webhook secret was updated.",
-          "type": "string"
+          "type": "string",
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:534]: This should use date-time format.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 1 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:232]: We need to pass `body` into our delete operation. As custom headers are not supported this is the suggested way I got during Office Hours.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 2 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:432]: This is OAuth state, which may be an arbitrary value. From GitHub definition: 

> An unguessable random string. It is used to protect against cross-site request forgery attacks.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 5]
```diff
@@ -700,6 +792,42 @@
...
+          "description": "State of the pull request",
+          "type": "object"
+        }
+      }
+    },
+    "PullRequestState": {
+      "description": "Status of the pull request.",
+      "enum": [
+        "Open",
+        "Closed"
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:816]: Only two states for PR?

Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:816]: Yes, as we're not interested in all other states. For us it is either "in flight" or "closed".

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 6]
```diff
@@ -163,6 +163,10 @@
               "$ref": "#/definitions/SourceControl"
             }
           },
+          "404": {
+            "description": "Not Found",
+            "x-ms-error-response": true
+          },
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:169]: This response code is not required in swagger: [https://github.com/Azure/azure-openapi-validator/blob/main/docs/no-error-code-responses.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/no-error-code-responses.md).

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 7]
```diff
@@ -240,29 +255,14 @@
...
-            "schema": {
-              "$ref": "#/definitions/SourceControl"
-            }
-          },
-          "201": {
-            "description": "Created",
-            "schema": {
-              "$ref": "#/definitions/SourceControl"
-            }
+            "description": "OK"
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:262]: 200 Response is never without response schema.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 8]
```diff
@@ -240,29 +255,14 @@
...
           },
-          "default": {
-            "description": "Error response describing why the operation failed.",
-            "schema": {
-              "$ref": "../../../common/2.0/types.json#/definitions/CloudError"
-            }
+          "204": {
+            "description": "No Content"
           }
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:266]: Why both 200 and 204 responses? It should have either of them, not both.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 9]
```diff
@@ -240,29 +255,14 @@
           },
           {
             "$ref": "#/parameters/SourceControlIdParameter"
-          },
-          {
-            "$ref": "#/parameters/SourceControl"
           }
         ],
         "responses": {
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:260]: Missing default response.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 10]
```diff
@@ -404,8 +417,41 @@
       ],
       "type": "object"
     },
+    "RepositoryAccess": {
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:420]: Schema should have a description or title.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 8 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:266]: We return `Warning`s as part of `200` response. `204` will not contain any warnings.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 10 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:425]: Fixed.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 9 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:256]: Fixed.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 7 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json]: Fixed.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 4 above)
Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:544]: Fixed.

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json (hunk 1 above)
Comment (visingla-ms) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:228]: There is no body in request payload now, as I can see. Should this be Delete operation now or is there any reason for it to be a POST Action?

File: specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json [hunk 11]
```diff
@@ -213,18 +222,20 @@
             }
           }
         }
-      },
-      "put": {
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.OperationalInsights/workspaces/{workspaceName}/providers/Microsoft.SecurityInsights/sourcecontrols/{sourceControlId}/delete": {
```
Comment (visingla-ms) [inline specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:227]: There is no body in request payload now, as I can see. Should this be Delete operation now or is there any reason for it to be a POST Action?

Comment (zjklee) [reply specification/securityinsights/resource-manager/Microsoft.SecurityInsights/preview/2023-06-01-preview/SourceControls.json:227]: Good catch, thanks. Added definition of the body.

**Comment by keryul** [comment]
@qiaozha please review the JS breaking change
@msyyc please review the Python breaking change

**Comment by zjklee** [comment]
@qiaozha, can you please have a look?

**Review by ankhyk — APPROVED** [review]

**Comment by zjklee** [comment]
@keryul , can it be merged now?

**Pull request merged**